### PR TITLE
Gpu mode

### PR DESCRIPTION
**Pull request by MChoi-git**
This PR investigates GPU and CPU mode for hooks, which allow the user to specify the location where activation tensors are cached. CPU mode minimizes GPU memory usage by caching activations in CPU RAM, at the cost of non-overlapped tensor D2H transfers. GPU mode reduces latency by caching activations in GPU DRAM, at the cost of potentially OOMing.

### CODE REVIEW

File: flex_model/core/hook_function.py [hunk 1]
```diff
@@ -159,7 +141,7 @@ def __init__(
         module_name: str,
         expected_shape: Tuple[Optional[int], ...],
         editing_function: Optional[Callable] = None,
-        hook_type: str = "forward",
+        unpack_idx: int = 0,
```
Comment (MChoi-git) [inline flex_model/core/hook_function.py:144]: Missing doc param.

File: flex_model/core/hook_function.py [hunk 2]
```diff
@@ -231,35 +203,58 @@ def _unpack_layer_outputs(
...
+        def _offload_tensor_to_cpu(activation: Tensor) -> None:
+            assert self._shared_state.output_ptr is not None
 
-        :param Tensor activation: Full activation tensor to save.
+            if not using_torch_dist() or (using_act_dist() and in_pp_group()):
+                # Tensor.to() with non_blockin=True operates on shared memory
+                # buffers automatically.
+                self._shared_state.output_ptr[
+                    self.module_name
+                ] = activation.detach().to("cpu", non_blocking=True)
```
Comment (MChoi-git) [inline flex_model/core/hook_function.py:237]: Change output dict from `Dict[str, Tensor] -> Dict[str, List[Tensor]]` since each module can have multiple hooks, similar in format to `FlexModel.hook_functions` collection.

File: flex_model/core/hook_function.py [hunk 3]
```diff
@@ -231,35 +203,58 @@ def _unpack_layer_outputs(
...
-            is bound the :class:`HookFunction` instance to save to.
-        """
-        assert self._output_ptr is not None
-        dumped_tensor = activation.detach().cpu()
-        self._output_ptr[self.module_name] = dumped_tensor
+        def _offload_tensor_to_gpu(activation: Tensor) -> None:
+            assert self._shared_state.output_ptr is not None
+
+            if not using_torch_dist() or (using_act_dist() and in_pp_group()):
+                self._shared_state.output_ptr[self.module_name] = activation.detach()
```
Comment (MChoi-git) [inline flex_model/core/hook_function.py:243]: call `clone()` after `detach()`.

File: flex_model/core/hook_function.py [hunk 4]
```diff
@@ -231,35 +203,58 @@ def _unpack_layer_outputs(
...
 
-    def _parse_tensor(self, tensor: Tensor) -> None:
+        mode_to_fn = {
+            "CPU": _offload_tensor_to_cpu,
+            "GPU": _offload_tensor_to_gpu,
+        }
```
Comment (MChoi-git) [inline flex_model/core/hook_function.py:248]: Make this a class attribute

File: flex_model/core/hook_function.py [hunk 5]
```diff
@@ -296,96 +291,98 @@ def _dispatch_hook_function(
...
         logger.debug(f"*{self.module_name}: Hook function activated*")
 
-        handle_fn = self._hook_type_to_handle_map[self.hook_type]
+        hook_type_to_impl_fn = {
+            "forward": self._forward_hook_impl,
+            "full_backward": self._full_backward_hook_impl,
+            "tensor": self._tensor_hook_impl,
+            "forward_pre": self._forward_pre_hook_impl,
+            "full_backward_pre": self._full_backward_pre_hook_impl,
+        }
```
Comment (MChoi-git) [inline flex_model/core/hook_function.py:302]: Make this also a class attribute. See if we can condense all of these mappings.

File: flex_model/core/hook_function.py [hunk 6]
```diff
@@ -296,96 +291,98 @@ def _dispatch_hook_function(
...
-        # Poplate the collection and dispersion functions.
-        if self._collect is None or self._disperse is None:
-            self._parse_tensor(tensor)
+        # Concretize functions.
+        if (
+            self._collect is None
+            or self._disperse is None
+            or self._edit is None
+            or self._offload is None
+        ):
```
Comment (MChoi-git) [inline flex_model/core/hook_function.py:364]: There should be a check to make sure these are either 1. All `None` or 2. All `callable`. Any mixed states could lead to silent failures and should be explicitly invalid.

File: flex_model/core/wrapper.py [hunk 7]
```diff
@@ -1,6 +1,12 @@
+import asyncio
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:1]: unused

File: flex_model/core/wrapper.py [hunk 8]
```diff
@@ -32,6 +40,48 @@
 logger = logging.getLogger(__name__)
 
 
+_OFFLOAD_MODES: Set[str] = {
+    "CPU",
+    "GPU",
+}
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:46]: Move into `FlexModel` as an attribute, or get rid of it and try to reuse the modes defined in `HookFunction`.

File: flex_model/core/wrapper.py [hunk 9]
```diff
@@ -32,6 +40,48 @@
...
+def get_module(root_module, tgt_module_name):
+    submodule_names = tgt_module_name.split(".")
+    # Can access submodules by `a.b`, but some modules require `a[b]`.
+    return functools.reduce(
+        lambda carry, n: getattr(carry, n)
+        if not isinstance(carry, (nn.ModuleList, nn.Sequential))
+        else carry[int(n)],
+        submodule_names,
+        root_module,
+    )
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:58]: Make private and type check

File: flex_model/core/wrapper.py [hunk 10]
```diff
@@ -32,6 +40,48 @@
...
+    # Clear distributed states.
+    if dist.distributed_backend_is_initialized():
+        if dist.activation_parallel_is_initialized():
+            dist.destroy_activation_parallel()
+        dist.destroy_distributed_backend()
+
+
 class FlexModel(nn.Module):
     """Wraps a Pytorch :code:`nn.Module` to provide an interface for various
     model-surgery techniques. Most importantly, allows registration of
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:87]: Should update docstrings for new parameters.

File: flex_model/core/wrapper.py [hunk 11]
```diff
@@ -133,46 +181,162 @@ def __init__(
...
-            to register.
-        """
+        # Initialize mappings.
+        self._hook_type_to_pt_attr = {
+            "forward": "register_forward_hook",
+            "full_backward": "register_full_backward_hook",
+            "tensor": "register_hook",
+            "forward_pre": "register_forward_pre_hook",
+            "backward_pre": "register_full_backward_pre_hook",
+        }
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:218]: This is also defined in `HookFunction`. There are many mappings defined in both classes, maaybe find a way to unify these for better code reuse

File: flex_model/core/wrapper.py [hunk 12]
```diff
@@ -133,46 +181,162 @@ def __init__(
...
+
+    @make_group.register
+    def _(self, group_constructor: str):
+        # TODO: Pattern match module names for already-registered hook functions.
+        raise NotImplementedError
+
+    @make_group.register
+    def _(self, group_constructor: list, group_alias: str):
+        # TODO: Construct group explicitly using hook functions.
+        raise NotImplementedError
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:241]: Unfinished this PR, make private

File: flex_model/core/wrapper.py [hunk 13]
```diff
@@ -133,46 +181,162 @@ def __init__(
...
+        # Set private HF state.
+        assert hook_function._shared_state is None
+        assert hook_function._hook_type is None
+        hook_function._shared_state = self._shared_state
+        hook_function._hook_type = hook_type
+
+        # Pass to registration impl.
+        self._register_hook_impl(hook_function)
+
+    def register_forward_hook(self, hook_function: HookFunction, lazy=False) -> None:
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:280]: Get rid of `lazy` in signatures, unused.

File: flex_model/core/wrapper.py [hunk 14]
```diff
@@ -133,46 +181,162 @@ def __init__(
...
-        hook_function._output_ptr = self.output_ptr
-        hook_function.save_ctx = self.save_ctx
-        hook_function.modules = self.trainable_modules
-        self.hook_functions[hook_function.module_name] = hook_function
+            # Other ranks reset their collections for the next forward pass.
+            else:
+                self.output_ptr = {}
+
+    def forward(self, *args, _group: str = "all", **kwargs):
+        self._enable_group_hooks(_group)
```
Comment (MChoi-git) [inline flex_model/core/wrapper.py:332]: Future `TODO`: Provide an interable of all group names and run the union of them.

File: profiling/profile_hooks.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+    # Construct experiment setup functions and create profile folders.
+    manager = ExperimentNetworkManager()
+    experiments = manager.named_experiments
+
+    prefix = ""
+    if args.single_gpu_only:
+        prefix = "single_gpu"
+    elif args.multi_gpu_only:
+        prefix = "multi_gpu"
+    experiments = [getattr(manager, e) for e in experiments if e.startswith(prefix)]
```
Comment (MChoi-git) [inline profiling/profile_hooks.py:145]: This can probably be pushed inside `ExperimentNetworkManager`, where we just request experiment handles via prefix-matching.

File: profiling/profile_hooks.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+
+    # Run benchmarks for each experiment, sweeping over parameters.
+    for model_dim in args.model_dim_sweep:
+        for dtype in args.dtype_sweep:
+            # Setup inputs and spoof config.
+            inputs = torch.randn(
+                num_steps, args.batch_size, model_dim, dtype=dtype
+            ).cuda()
+
+            spoof_config = spoof_megatron_config(dtype)
```
Comment (MChoi-git) [inline profiling/profile_hooks.py:170]: Note why spoofing the megatron-lm config is necessary.

File: profiling/profile_hooks.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+
+            for exp in experiments:
+                exp_name = exp.__name__
+                if args.profile_force and exp_name != args.profile_force:
```
Comment (MChoi-git) [inline profiling/profile_hooks.py:174]: change `profile_force` -> `profile_force_exp` to be more descriptive.

File: profiling/test.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
+import argparse
```
Comment (MChoi-git) [inline profiling/test.py:1]: file can be removed

File: profiling/utils.py [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+    ColumnParallelLinear,
+    RowParallelLinear,
+    model_parallel_cuda_manual_seed,
+)
+from torch.profiler import profile
+
+from flex_model.core import FlexModel, HookFunction
+
+
+class TestNetwork(nn.Module):
```
Comment (MChoi-git) [inline profiling/utils.py:26]: Make a note in a docstring about the configuration options of this NN.

File: profiling/utils.py [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+        gather_fn = all_gather
+        scatter_fn = scatter
+
+    elif isinstance(self, RowParallelLinear):
+        gather_fn = lambda x: x
+        scatter_fn = lambda x: x
+
+    else:
+        gather_fn = lambda x: x
+        scatter_fn = lambda x: x
```
Comment (MChoi-git) [inline profiling/utils.py:144]: This option should be `isinstance(self, nn.Linear)` and we have a fallback `else` which raises an implementation error.

File: profiling/utils.py (hunk 19 above)
Comment (MChoi-git) [inline profiling/utils.py:26]: Also `TestNetwork` should be deprecated in the next testing PR in favour of Megatron-LM models for working with 3-D sharding schemes and training/finetuning workloads.

File: profiling/utils.py [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+            network.register_forward_hook(
+                HookFunction(module_name=n, expected_shape=(None, model_dim),)
+            )
+        return network
+
+
+def init_megatron_dist(args):
+    """Initialize Megatron-LM parallel state."""
+    os.environ["NCCL_IB_DISABLE"] = "1"
+    initialize_distributed()
```
Comment (MChoi-git) [inline profiling/utils.py:332]: there are lots of distributed backends being enabled at this point, specify this is torch distributed.

File: profiling/utils.py [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+    initialize_model_parallel(args.tp_size)
```
Comment (MChoi-git) [inline profiling/utils.py:333]: maybe alias this as `initialize_megatronlm_model_parallel`

File: tests/multi_gpu/core/test_fairscale_megatron.py [hunk 23]
```diff
@@ -137,7 +133,69 @@ def test_forward_hooks_FairscaleLayers():
...
+            output_dict["row_parallel_linear"], output_dict["row_linear"], atol=1e-7
+        )
+        logger.info("Tests successful.")
+
+    Utils.destroy_model_parallel()
+
+
+@register_fairscale_megatron_test
+def test_gpu_offload_FairscaleLayers():
+    Utils.initialize_model_parallel(2, 1, 2)
```
Comment (MChoi-git) [inline tests/multi_gpu/core/test_fairscale_megatron.py:162]: Maybe create a utility similar to `pytest.mark.parametrize(...)` for usage with the `submitit` multi-gpu tests.

**Review by MChoi-git — COMMENTED** [review]
Added comments/revisions that need to be addresses before merging

**Review by adil-a — COMMENTED** [review]
Looks good overall. Left some comments regarding style consistency + other small things.
